design/Extend-VolumePolicies-to-support-more-actions.md

```diff
@@ -0,0 +1,176 @@
+introducing the VolumePolicies feature was to improve the overall user experience when performing backup operations
+for volume resources, the feature enables users to group volumes according the `conditions` (criteria) specified and
+also lets you specify the `action` that velero needs to take for these grouped volumes during the backup operation.
+The limitation being that currently `VolumePolicies` only supports `skip` as an action, We want to extend the `action`
+functionality to support more usable options like `fs-backup` (File system backup) and `snapshot` (VolumeSnapshots).
+
+## Goals
+- Extending the VolumePolicies to support more actions like `fs-backup` (File system backup) and `snapshot` (VolumeSnapshots).
+- Improve user experience when backing up Volumes via Velero
+
+## Non-Goals
+- No changes to existing approaches to opt-in/opt-out annotations for volumes
+- No changes to existing `VolumePolicies` functionalities
+
+
+## Use-cases/Scenarios
+
+**Use-case 1:**
+- A user wants to use `snapshot` (volumesnapshots) backup option for all the csi supported volumes and `fs-backup` for the rest of the volumes.
+- Currently, velero supports this use-case but the user experience is not that great.
+- The user will have to individually annotate the volume mounting pod with the annotation "backup.velero.io/backup-volumes" for `fs-backup`
+- This becomes cumbersome at scale.
+- Using `VolumePolicies`, the user can just specify 2 simple `VolumePolicies` like for csi supported volumes as `snapshot` action and rest can be backed up`fs-backup` action:
+```yaml
+version: v1
+volumePolicies:
+- conditions:
+    storageClass:
+    - gp2
+  action:
+    type: snapshot
+- conditions: {}
+  action:
+    type: fs-backup
+```
+
+**Use-case 2:**
+- A user wants to use `fs-backup` for nfs volumes pertaining to a particular server
+- In such a scenario the user can just specify a `VolumePolicy` like:
+```yaml
+version: v1
+volumePolicies:
+- conditions:
+    nfs:
+      server: 192.168.200.90
+  action:
+    type: fs-backup
+```
+## High-Level Design
+- When the VolumePolicy action is set as `fs-backup` the backup workflow modifications would be:
+  - We call [backupItem() -> backupItemInternal()](https://github.com/vmware-tanzu/velero/blob/main/pkg/backup/item_backupper.go#L95) on all the items that are to be backed up
+  - Here when we encounter [Pod as an item ](https://github.com/vmware-tanzu/velero/blob/main/pkg/backup/item_backupper.go#L195)
+  - We will have to modify the backup workflow to account for the `fs-backup` VolumePolicy action
+
+**Note:** `fs-backup` via VolumePolicy and `fs-backup` via annotation based opt-in/opt-out approach would be mutually exclusive.
+
+- When the VolumePolicy action is set as `snapshot` the backup workflow modifications would be:
+  - Once again, We call [backupItem() -> backupItemInternal()](https://github.com/vmware-tanzu/velero/blob/main/pkg/backup/item_backupper.go#L95) on all the items that are to be backed up
```

Comment:
The code logic in `backupItemInternal ` is quite disorganized, with three methods—CSI snapshot, fs-backup, and native-snapshot—all intertwined. Have you considered reorganizing them, and creating separate functions or modules for each, to make the logic clearer for each respective backup method? @shubham-pampattiwar

Reply:
@qiuming-best Even if we convert to modules per backup method for volumes, the code path would remain the same right ?
like if object encountered is Pod, check for FSBackup option 
if PV object encountered then check for snapshot option
Additionally, I wanted to make changes without affecting the existing workflow that is properly working.

Reply:
I understand it's too risky to rework the existing backup workflow, but I wish to suggest we introduce one data structure or struct `volumeHelper` to help we decide what to do with a PV or PVC.

Ideally, we should have `volumeHelper.GetAction(PV)` to tell us if we wanna take snapshot or use fs-backup or skip the volume, so volumeHelper should consolidate the content in volumepolicy and backup spec.